design-documents/graph-ql/coverage/negotiableQuotes.graphqls

```diff
@@ -0,0 +1,361 @@
+    shipping_amount: Float @doc(description: "Shipping amount")
+    base_shipping_amount: Float @doc(description: "Base shipping amount")
+    shipping_discount_amount: Float @doc(description: "Shipping discount amount")
+    tax_amount: Float @doc(description: "Tax amount")
+    base_tax_amount: Float @doc(description: "base_tax_amount")
+    shipping_tax_amount: Float @doc(description: "shipping_tax_amount")
+    subtotal_incl_tax: Float @doc(description: "subtotal_incl_tax")
+    shipping_incl_tax: Float @doc(description: "shipping_incl_tax")
+    base_shipping_incl_tax: Float @doc(description: "base_shipping_incl_tax")
+    base_currency_code: String @doc(description: "base_currency_code")
+    quote_currency_code: String @doc(description: "quote_currency_code")
+    items_qty: Int @doc(description: "items_qty")
+    items: [ItemsOutput] @doc(description: "Items details")
+    total_segments: [TotalSegmentOutput] @doc(description: "Total segments")
+}
+
+type ItemsOutput {
+    item_id: ID @doc(description: "Item ID")
+    price: Float @doc(description: "Price")
+    base_price: Float @doc(description: "Base price")
+    qty: Int @doc(description: "Quantity")
+    row_total: Float @doc(description: "Row total")
+    base_row_total: Float @doc(description: "Base row total")
+    row_total_with_discount: Float @doc(description: "Row total with discount")
+    tax_amount: Float @doc(description: "Tax amount")
+    base_tax_amount: Float @doc(description: "Base tax amount")
+    tax_percent: Float @doc(description: "Tax percent")
+    discount_amount: Float @doc(description: "Discount amount")
+    base_discount_amount: Float @doc(description: "Base discount amount")
+    discount_percent: Float @doc(description: "Discount percent")
+    price_incl_tax: Float @doc(description: "Price incl. tax")
+    base_price_incl_tax: Float @doc(description: "Base price incl. tax")
+    row_total_incl_tax: Float @doc(description: "Row total incl. tax")
+    base_row_total_incl_tax: Float @doc(description: "Base row total incl. tax")
+    name: String @doc(description: "Name")
+}
+
+type TotalSegmentOutput {
+    code: String @doc(description: "Segment code")
+    title: String @doc(description: "Segment title")
+    value: Float @doc(description: "Segment value")
+}
+
+type TotalInfo {
+    totals: TotalsOutput! @doc(description: "List of order otals for negotiable quote")
+}
+
+input UpdatedPricesInput {
+    quote_ids: [String]! @doc(description: "IDs of the updated quotes")
+}
+
+input NegotiableQuoteRequestInput {
+    quote_id: String! @doc(description: "Quote hash")
+    quote_name: String @doc(description: "Quote name")
+    comment: String @doc(description: "Comment")
+    files: [FileInput] @doc(description: "Attached files")
+}
+
+type RequestQuoteOutput {
+    negotiable_quote:NegotiableQuote!
+}
+
+input FileInput @doc(description: "The list of file attachment codes") {
```

Comment:
I'm a bit worried about doing file uploads through GraphQL. People certainly do it, but it may not be ideal for us.

It's not that big of a deal while GraphQL is implemented in the monolith.

But, as we explore the possibility of decoupling GraphQL from the PHP application, uploads introduce some problems:

1.  The GraphQL server can't begin processing any of the request (including parsing the mutation) until the whole file is uploaded
2.  The GraphQL server won't know enough about the request (because of 1) to start calling the backing service, so the whole file needs to be written to storage or buffered in memory, and can't just be streamed to its destination
3. Sending binary files inside GraphQL types means the whole file runs through the GraphQL parsing algorithm, which is unnecessary synchronous work

Reply:
@mbrinton01 @nrkapoor  what do you want to do here?

Reply:
I think this is something we will need to support at some point. here is a pointer if it helps. https://medium.com/@enespalaz/file-upload-with-graphql-9a4927775ef7

Reply:
@nrkapoor Not saying this is something we shouldn't support, just saying I don't think this is the way to do it (passing binary files through the GraphQL server).

That document describes that it is possible from a technical perspective. But the document doesn't address any of the outlined concerns.

Reply:
Created internal ticket (ECP-750) for spike on uploads in general for GraphQL. Up to @nrkapoor whether he wants to proceed with this part of the schema for now or just defer.